io/include/detray/io/common/geometry_writer.hpp

```diff
@@ -245,6 +217,66 @@ class geometry_writer : public writer_interface<detector_t> {
                 typename material_group_t::value_type>(index);
         }
     };
+
+    /// Retrieve @c acc_links_payload from surface_tore collection
+    struct get_acc_link_payload {
+        template <typename acc_group_t, typename index_t>
+        constexpr inline auto operator()(const acc_group_t&,
+                                         const index_t& index) const {
+
+            using accelerator_t = typename acc_group_t::value_type;
+
+            if constexpr (detail::is_grid_v<accelerator_t>) {
+                constexpr auto id{
+                    detail::get_grid_id(accelerator_t::local_frame())};
+                return geometry_writer<detector_t>::serialize(id, index);
+            } else {
+                // This functor is only called for accelerator data structures
+                // that are not 'brute force'
+                return geometry_writer<detector_t>::serialize(
+                    acc_links_payload::acc_type::unknown, index);
+            }
+        }
+    };
+
+    /// Infer the IO shape id from the shape type
+    template <typename shape_t>
+    static constexpr io::detail::mask_shape get_shape_id() {
+
+        // Find the correct shape IO id;
+        if constexpr (std::is_same_v<shape_t, rectangle2D<>>) {
+            return io::detail::mask_shape::rectangle2;
+        } else if constexpr (std::is_same_v<shape_t, trapezoid2D<>>) {
+            return io::detail::mask_shape::trapezoid2;
+        } else if constexpr (std::is_same_v<shape_t, cylinder2D<>>) {
+            return io::detail::mask_shape::cylinder2;
+        } else if constexpr (std::is_same_v<
+                                 shape_t,
+                                 cylinder2D<false,
+                                            cylinder_portal_intersector>>) {
+            return io::detail::mask_shape::portal_cylinder2;
+        } else if constexpr (std::is_same_v<shape_t, ring2D<>>) {
+            return io::detail::mask_shape::ring2;
+        } else if constexpr (std::is_same_v<shape_t, annulus2D<>>) {
+            return io::detail::mask_shape::annulus2;
+        } else if constexpr (std::is_same_v<shape_t, line<true>>) {
+            return io::detail::mask_shape::cell_wire;
+        } else if constexpr (std::is_same_v<shape_t, line<false>>) {
+            return io::detail::mask_shape::straw_wire;
+        } else if constexpr (std::is_same_v<shape_t, single3D<0>>) {
+            return io::detail::mask_shape::single1;
+        } else if constexpr (std::is_same_v<shape_t, single3D<1>>) {
+            return io::detail::mask_shape::single2;
+        } else if constexpr (std::is_same_v<shape_t, single3D<2>>) {
+            return io::detail::mask_shape::single3;
+        } else if constexpr (std::is_same_v<shape_t, cuboid3D<>>) {
+            return io::detail::mask_shape::cuboid3;
+        } else if constexpr (std::is_same_v<shape_t, cylinder3D>) {
+            return io::detail::mask_shape::cylinder3;
```

Comment:
What I mean here is - if we have somewhere a tuple with all shape types, on could use this to macht and return ...

Reply:
Done